### PR TITLE
Gracefully halt downstream jobs on CircleCI when parent jobs are skipped

### PR DESCRIPTION
**Pull request by danielrozenberg**
This PR uses [CircleCI's workspaces](https://circleci.com/blog/persisting-data-in-workflows-when-to-use-caching-artifacts-and-workspaces/) to indicate to downstream jobs that they should not be executed; this happens before checking out the code, so a job that is going to be skipped completes within ~15-20 seconds instead of ~2-3 minutes

### CODE REVIEW

**Comment by danielrozenberg** [comment]
@rsimha it's not exactly easy to test this on real-world builds because different PRs would have different build targets. Instead of calling `circleci-agent step halt`, we can just log this for now and then run through X days of PR builds to see whether this works. wdyt?

File: .circleci/config.yml [hunk 1]
```diff
@@ -143,36 +170,49 @@ jobs:
...
       - fail_fast
+      - teardown_vm
   'Visual Diff Tests':
     executor:
       name: amphtml-large-executor
     steps:
+      # NOTE: deliberately not running the `maybe_gracefully_halt` step because
+      # visual diff tests need to at least run with --empty when skipped.
+      # Instead we must create an empty workspace to prevent `teardown_vm` from
+      # crashing.
```
Comment (rsimha) [inline .circleci/config.yml:181]: I recognize that this special case exists because Percy is a required check. However, this looks like a detail of implementation of visual testing, and I think it should live in nodejs code and not in the main CircleCI config file.

Alternative suggestion: Move the call to `timedExecOrDie('amp visual-diff --empty');` from `pr-check/visual-diff-tests.js` to this section of `pr-check/nomodule-build.js` (say, after line 70). 

https://github.com/ampproject/amphtml/blob/20813ccd2bcdc7c80defb601081c7509ef2455c0/build-system/pr-check/nomodule-build.js#L69-L75

With this change, we can gracefully halt the `Visual Diff Tests` job without having to pay the additional task of VM setup just to create a blank Percy build, and without this duplicate `Create Workspace` step.

**Review by rsimha — COMMENTED** [review]
I'm excited by the potential CPU and time savings in this PR! Two comments below, lemme know what you think.

File: .circleci/config.yml [hunk 2]
```diff
@@ -47,8 +47,18 @@ commands:
       - restore_cache:
           name: 'Restore Merge Commit'
           key: git-{{ .Branch }}-{{ .Revision }}
+  maybe_gracefully_halt:
+    steps:
+      - attach_workspace:
+          at: /tmp
+      - run:
+          name: 'Maybe Gracefully Halt'
+          command: if [[ -f "/tmp/workspace/CI_GRACEFULLY_HALT" ]]; then echo "Gracefully halting this job."; circleci-agent step halt; fi
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply .circleci/config.yml:56]: Unfortunately that would mean we'd have to check out the repository first - which is the thing we are optimizing out here :)

We could do that and use `curl` in the same way that the `Compute Merge Commit` does, but this seems like an overkill just for a single "if" command

File: .circleci/config.yml (hunk 1 above)
Comment (danielrozenberg) [reply .circleci/config.yml:181]: Makes sense! I'll do that

File: .circleci/config.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rsimha) [reply .circleci/config.yml:56]: Aaah, excellent point. Carry on.

File: .circleci/config.yml (hunk 1 above)
Comment (danielrozenberg) [reply .circleci/config.yml:181]: Done

**Review by rsimha — COMMENTED** [review]
Almost there! Two comments below, containing a few suggestions.

Also, I looked at the CircleCI builds, this PR is converting this...

image

... to this...

image

👏 👏 👏

File: build-system/pr-check/utils.js [hunk 3]
```diff
@@ -108,11 +109,12 @@ function printChangeSummary() {
  * @param {string} jobName
  * @param {string} skipReason
  */
-function printSkipMessage(jobName, skipReason) {
+function skipFollowupJobs(jobName, skipReason) {
   const loggingPrefix = getLoggingPrefix();
   logWithoutTimestamp(
     `${loggingPrefix} Skipping ${cyan(jobName)} because ${skipReason}.`
   );
+  fs.closeSync(fs.openSync('/tmp/workspace/.CI_GRACEFULLY_HALT', 'w'));
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/utils.js:117]: Done

File: build-system/pr-check/nomodule-build.js [hunk 4]
```diff
@@ -68,10 +68,15 @@ async function prBuildWorkflow() {
...
-    printSkipMessage(
+    skipFollowupJobs(
       jobName,
       'this PR does not affect the runtime, integration tests, end-to-end tests, or visual diff tests'
     );
+
+    // Special case for visual diffs - Percy is a required check and must pass,
+    // but we just called `skipFollowupJobs` so the Visual Diffs job will not
+    // run. Instead, we create an empty, passing check on Percy here.
+    timedExecOrDie('amp visual-diff --empty');
```
[the remark this answers is not in the document]
Comment (danielrozenberg) [reply build-system/pr-check/nomodule-build.js]: Done.

File: build-system/common/ci.js [hunk 5]
```diff
@@ -225,6 +226,17 @@ function ciBuildSha() {
   return isPullRequestBuild() ? ciPullRequestSha() : ciCommitSha();
 }
 
+/**
+ * Signal to dependent jobs that they should be skipped.
+ *
+ * Currently only relevant for CircleCI builds.
+ */
+function signalGracefulHalt() {
```
Comment (rsimha) [inline build-system/common/ci.js:234]: Since `ci.js` only provides existing state from the environment, I think this function should live in `utils.js`.

**Review by rsimha — APPROVED** [review]
Yay! LGTM after one last change.

**Comment by danielrozenberg** [comment]
Some numbers - checked using CircleCI's API how long this vs other pull requests take to run:

Total VM time (sum total of all jobs that have been parallelized)
```
pull/* (63 workflows)
Median: 2:09:11
Slowest quartile: 2:13:07
Slowest 5%: 2:18:59

pull/33620 (10 workflows)
Median: 2:06:28
Slowest quartile: 2:07:11
Slowest 5%: 2:08:45
```

Wall time (how long people have to wait for workflows to complete):
```
pull/33620 (10 workflows)
Median: 0:21:21
Slowest quartile: 0:21:45
Slowest 5%: 0:24:03

pull/* (63 workflows)
Median: 0:20:50
Slowest quartile: 0:21:28
Slowest 5%: 0:22:57
```

Note that "other PRs" include those PRs that take less time to run when they only change e.g., docs. If I drop the PRs that take less than 15 minutes to complete their workflow (a rough estimate) the results change to:

```
pull/33620 (10 workflows)
Median: 0:21:21
Slowest quartile: 0:21:45
Slowest 5%: 0:24:03

pull/* (56 workflows)
Median: 0:21:06
Slowest quartile: 0:21:32
Slowest 5%: 0:23:21
```

```
pull/33620 (10 workflows)
Median: 2:06:28
Slowest quartile: 2:07:11
Slowest 5%: 2:08:45

pull/* (56 workflows)
Median: 2:09:57
Slowest quartile: 2:13:36
Slowest 5%: 2:18:59
```

TL;DR: wall time is _increased_ by ~15 seconds for 50% of jobs, but VM time _decreases_ by ~3 minutes for 50% of jobs and ~6 minutes for 25% of jobs

File: build-system/common/ci.js (hunk 5 above)
Comment (danielrozenberg) [reply build-system/common/ci.js]: Done

**Comment by rsimha** [comment]
> wall time is _increased_ by ~15 seconds for 50%

My suspicion is that this is due to queueing (captured by a slight variance in the time taken by the `Spin up environment` step). And my gut feeling is that in the long run, this PR will decrease both wall time and VM time because it immediately frees up VMs that don't have any work to do.

File: .circleci/config.yml (hunk 1 above)
Comment (rsimha) [reply .circleci/config.yml]: @danielrozenberg We forgot one thing. Skipping a Percy build also needs the token. See [this broken build](https://app.circleci.com/pipelines/github/ampproject/amphtml/6229/workflows/0d129b3f-f949-42df-bb67-e2ec2dbdbc69/jobs/88512/parallel-runs/0/steps/0-111).

We'll need to insert this code:

https://github.com/ampproject/amphtml/blob/d71ab5009757675e476b7f3ea26bfd97ef1d8d3d/build-system/pr-check/visual-diff-tests.js#L40

PR incoming.
